app/ante/ante.go

```diff
@@ -33,66 +32,75 @@ type HandlerOptions struct {
+	// Define a custom free lane fee checker.
 	freeLaneFeeChecker := func(ctx sdk.Context, tx sdk.Tx) (sdk.Coins, int64, error) {
-		// skip fee checker if the tx is free lane tx.
 		if !options.FreeLane.Match(ctx, tx) {
 			return txFeeChecker(ctx, tx)
 		}
-
-		// return fee without fee check
 		feeTx, ok := tx.(sdk.FeeTx)
 		if !ok {
 			return nil, 0, errors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
 		}
-
 		return feeTx.GetFee(), 1 /* FIFO */, nil
 	}
 
-	anteDecorators := []sdk.AnteDecorator{
+	// Create the AnteDecorators sequence.
+	anteDecorators := buildAnteDecorators(options, freeLaneFeeChecker)
+
+	// Chain the AnteDecorators to construct the AnteHandler.
+	return sdk.ChainAnteDecorators(anteDecorators...), nil
+}
+
+// validateHandlerOptions ensures all mandatory dependencies are provided.
+func validateHandlerOptions(options HandlerOptions) error {
+	if options.AccountKeeper == nil {
+		return errors.Wrap(sdkerrors.ErrLogic, "account keeper is required for ante builder")
+	}
+	if options.BankKeeper == nil {
+		return errors.Wrap(sdkerrors.ErrLogic, "bank keeper is required for ante builder")
+	}
+	if options.SignModeHandler == nil {
+		return errors.Wrap(sdkerrors.ErrLogic, "sign mode handler is required for ante builder")
+	}
+	return nil
+}
+
+// getTxFeeChecker returns the appropriate TxFeeChecker function.
+func getTxFeeChecker(options HandlerOptions) func(sdk.Context, sdk.Tx) (sdk.Coins, int64, error) {
+	if options.TxFeeChecker != nil {
+		return options.TxFeeChecker
+	}
+	return moveante.NewMempoolFeeChecker(options.MoveKeeper).CheckTxFeeWithMinGasPrices
+}
+
+// buildAnteDecorators constructs the list of AnteDecorators in the correct order.
+func buildAnteDecorators(options HandlerOptions, freeLaneFeeChecker func(sdk.Context, sdk.Tx) (sdk.Coins, int64, error)) []sdk.AnteDecorator {
+	return []sdk.AnteDecorator{
 		accnum.NewAccountNumberDecorator(options.AccountKeeper),
-		ante.NewSetUpContextDecorator(), // outermost AnteDecorator. SetUpContext must be called first
+		ante.NewSetUpContextDecorator(), // Must be the first decorator.
 		ante.NewExtensionOptionsDecorator(options.ExtensionOptionChecker),
 		moveante.NewGasPricesDecorator(),
 		ante.NewValidateBasicDecorator(),
 		ante.NewTxTimeoutHeightDecorator(),
 		ante.NewValidateMemoDecorator(options.AccountKeeper),
 		ante.NewConsumeGasForTxSizeDecorator(options.AccountKeeper),
 		ante.NewDeductFeeDecorator(options.AccountKeeper, options.BankKeeper, options.FeegrantKeeper, freeLaneFeeChecker),
-		// SetPubKeyDecorator must be called before all signature verification decorators
-		ante.NewSetPubKeyDecorator(options.AccountKeeper),
+		ante.NewSetPubKeyDecorator(options.AccountKeeper), // Must be called before signature verification.
 		ante.NewValidateSigCountDecorator(options.AccountKeeper),
-		ante.NewSigGasConsumeDecorator(options.AccountKeeper, sigGasConsumer),
+		ante.NewSigGasConsumeDecorator(options.AccountKeeper, sigverify.DefaultSigVerificationGasConsumer),
```

Comment:
can you change sigGasConsumer to retrieve from options first?

Reply:
updated in the last commit, Merry Christmas ~!